### ISSUE TITLE
EventEmitter.emit is not asynchronous in Node.prototype.receive

### ISSUE THREAD

**Issue by ozomer** [node-red/node-red#833]
EventEmitter.emit is not async.
This means that in [Node.js line 202](https://github.com/node-red/node-red/blob/392fd6fed3c56e7ca3b42c20de3118211f7535c1/red/runtime/nodes/Node.js#L202) the handlers of the "input" event are called immediately (i.e. `this.emit(...)` pauses until they finish).
One unwanted side effect is that function nodes' ["duration" metric](https://github.com/node-red/node-red/blob/392fd6fed3c56e7ca3b42c20de3118211f7535c1/nodes/core/core/80-function.js#L171-L181) measures not only the time it took the the current node to handle the message, but also the time it took the following nodes to handle the message.

There are two ways to solve this issue. One way is to replace `this.emit(...)` with:
`var self = this; setImmediate(function() { self.emit(...); });`
(or `process.nextTick` instead of `setImmediate`).

The change means that any call to `receive`, from anywhere in the code, does not wait for the receiving node to actually finish handling the message.

**Comment by knolleary** [comment]
Yes, we're aware of this. Not a change to make lightly as we need to ensure there are no unexpected side effects in what could lead to a change in node execution order for flows that have branches in. Whilst we've never made any statements on the order a flow executes when it branches, we do have to be a bit careful.

**Comment by ozomer** [comment]
You can add a function like `Node.async_send` that will call `setImmediate(function() { node.send(...); });`, and recommend people to use it when possible.

**Comment by ozomer** [comment]
You can also make it depend on some environment variable, like `process.env.NODE_RED_ASYNC_SEND`.

**Comment by knolleary** [comment]
We're not going to add a second function like `async_send` as there's no reason to confuse the api with essentially duplicate functions that leaves the node author having to understand the subtle differences between them.

We will evaluate the impact of changing the core behaviour to be async in the near future. I don't expect it to be a big deal - but as I said, we do want to be a little careful over it.

**Comment by btsimonh** [comment]
sorry, very late comment :) Please Please Please don't change the default behaviour;  it's a far better option to add a parameter to the node.send call....

**Comment by knolleary** [comment]
@btsimonh could you expand on why you wouldn't want the behaviour to change?

**Comment by btsimonh** [comment]
Hi Nick, 
I know it's not a great way of working, but I certainly have some flows where the order of execution of paths is important.
e.g. in one flow, I send the msg to one chain to measure widths of text which would be produced; when this returns, I use those widths to change the text size, and send the same msg to a second chain to actually render the text at the appropriate size to fit.
I also have some (older) naughty flows which modify the msg, and then use the mod in a previous node.

Another stupid example;
I have flows which write thousands of files.  I had to create a synchronous file writer, else nodejs just ate all the ram as I pumped files as fast as possible into the async file writer.  This flow now relies on the file writer returning when it's done.  Obviously the correct way to solve this would be to create an async file write which output a msg when it's done, and have a feedback loop...  but that's not how it's implemented at the moment.

The fact is that a general change would break many existing flows.

For consideration:
Add a config to a flow for the default behaviour; make the default for existing flows (or imported flows with no config) synchronous (or, as was), and the default for newly created flows Async.  Then no existing flows would be broken, and users would be encouraged to think and write for async.
(personally, I had a hard time changing my mindset from a std synchronous programming background, and even after 18 months, I'm still 'getting there' :) ).
